### ISSUE TITLE
Erroneous borrow error from 2018 edition's NLL migration mode.

### ISSUE THREAD

**Issue by nrc** [rust-lang/rust#52967]
I have:

```
pub fn infer_defaults(&mut self, project_dir: &Path) -> CargoResult<()> {
    ....
    match &mut self.target_dir {
            Inferrable::Specified(Some(ref mut path)) if path.is_relative() => {
                *path = project_dir.join(&path);
            }
            _ => {},
        }
    ...
}
```
(from https://github.com/rust-lang-nursery/rls/blob/master/src/config.rs#L238-L277)

And I get the following warning:

```
warning[E0502]: cannot borrow `_` as mutable because it is also borrowed as immutable
   --> src/config.rs:259:40
    |
257 |         match &mut self.target_dir {
    |               -------------------- immutable borrow occurs here
258 |             // We require an absolute path, so adjust a relative one if it's passed.
259 |             Inferrable::Specified(Some(ref mut path)) if path.is_relative() => {
    |                                        ^^^^^^^^^^^^ mutable borrow occurs here
...
262 |             _ => {},
    |             - borrow later used here
    |
    = warning: This error has been downgraded to a warning for backwards compatibility with previous releases.
            It represents potential unsoundness in your code.
            This warning will become a hard error in the future.
```

The warning is the same without the `&mut`.

It seems to me this code is fine and there should not be a warning.

cc @nikomatsakis @pnkfelix

**Comment by pnkfelix** [comment]
Smaller example ([playground](https://play.rust-lang.org/?gist=a0d6c843d2616d0cfd91a63c74a69eef&version=nightly&mode=debug&edition=2018)):

```rust
fn main() {
    let mut stuff = ("left", "right");
    match stuff {
        (ref mut left, _) if *left == "left" => { }
        _ => {}
    }
}
```

**Comment by pnkfelix** [comment]
Here's the good news: this is not an NLL bug, per se.

In particular, when we opt into NLL explicitly, the example works ([playground](https://play.rust-lang.org/?gist=32eaa41bc4f7fdb138a75def4a89e0a2&version=nightly&mode=debug&edition=2018)): 

```rust
#![feature(nll)]

fn main() {
    let mut stuff = ("left", "right");
    match stuff {
        (ref mut left, _) if *left == "left" => { }
        _ => {}
    }
}
```
There *is* a bug in how I linked the migration mode into the 2018 edition. Basically, its not toggling on enough switches, and one of the ones I missed is causing something to break with the new way we handle `match`.

**Comment by pnkfelix** [comment]
This issue strikes me as an argument that we should turn on something analogous to compare-mode=nll except specialized to either the migration mode or to the 2018 edition.

**Comment by pnkfelix** [comment]
In particular, you see the error with `-Z borrowck=mir`, but not with `-Z borrowck=mir -Z two-phase-borrows`. That's a pretty strong hint that I forgot to have edition=2008 imply two-phase-borrows.
